File: pascal/escape.pas
{$mode objfpc}
unit escape;

interface

const reset                     : string = '0';
const bold                      : string = '1';
const faint                     : string = '2';
const italic                    : string = '3';
const underline                 : string = '4';
const blink_slow                : string = '5';
const blink_rapid               : string = '6';
const reverse                   : string = '7';
const conceal                   : string = '8';
const crossout                  : string = '9';
const font_primary              : string = '10';
const font_alternate_1          : string = '11';
const font_alternate_2          : string = '12';
const font_alternate_3          : string = '13';
const font_alternate_4          : string = '14';
const font_alternate_5          : string = '15';
const font_alternate_6          : string = '16';
const font_alternate_7          : string = '17';
const font_alternate_8          : string = '18';
const font_alternate_9          : string = '19';
const fraktur                   : string = '20';
const bold_off                  : string = '21';
const underline_double          : string = '21';
const color_normal              : string = '22';
const italic_off                : string = '23';
const fraktur_off               : string = '23';
const underline_off             : string = '24';
const blink_off                 : string = '25';
const inverse_off               : string = '27';
const reveal                    : string = '28';
const crossout_off              : string = '29';
const fg_black                  : string = '30';
const fg_red                    : string = '31';
const fg_green                  : string = '32';
const fg_yellow                 : string = '33';
const fg_blue                   : string = '34';
const fg_magenta                : string = '35';
const fg_cyan                   : string = '36';
const fg_white                  : string = '37';
const fg                        : string = '38';
const fg_default                : string = '39';
const bg_black                  : string = '40';
const bg_red                    : string = '41';
const bg_green                  : string = '42';
const bg_yellow                 : string = '43';
const bg_blue                   : string = '44';
const bg_magenta                : string = '45';
const bg_cyan                   : string = '46';
const bg_white                  : string = '47';
const bg                        : string = '48';
const bg_default                : string = '49';
const frame                     : string = '51';
const encircle                  : string = '52';
const overline                  : string = '53';
const frame_off                 : string = '54';
const encircle_off              : string = '54';
const overline_off              : string = '55';
const ideogram_underline        : string = '60';
const ideogram_underline_double : string = '61';
const ideogram_overline         : string = '62';
const ideogram_overline_double  : string = '63';
const ideogram_stressmark       : string = '64';
const ideogram_off              : string = '65';
const fg_bright_black           : string = '90';
const fg_bright_red             : string = '91';
const fg_bright_green           : string = '92';
const fg_bright_yellow          : string = '93';
const fg_bright_blue            : string = '94';
const fg_bright_magenta         : string = '95';
const fg_bright_cyan            : string = '96';
const fg_bright_white           : string = '97';
const bg_bright_black           : string = '100';
const bg_bright_red             : string = '101';
const bg_bright_green           : string = '102';
const bg_bright_yellow          : string = '103';
const bg_bright_blue            : string = '104';
const bg_bright_magenta         : string = '105';
const bg_bright_cyan            : string = '106';
const bg_bright_white           : string = '107';

function ansify(codes : array of const; const msg : string) : string;

implementation
uses sysutils;

const ansi_begin : string = #27'[';
const ansi_end : string = 'm';
const ansi_separator : char = ';';

function codify(codes : array of const) : string;
const
    COLOR_MODE : array[1..3] of string = ('5', '5', '2');

var _color : array[1..3] of byte;
    _clind : 1..3 = 1;
    _catch : boolean = false;
    _param : string = '';
    i : word;

    function color_to_string(const color : array of byte) : string;
    var c : byte;
    begin
        color_to_string := '';
        for c in color do
            color_to_string += IntToStr(c) + ansi_separator;
        Delete(color_to_string, Length(color_to_string) - 1, 1);
    end;

    procedure flush_color;
    begin
        _param += COLOR_MODE[_clind] + ansi_separator + color_to_string(_color[1.._clind]);
    end;

begin
    for i := 0 to High(codes) do begin
        case codes[i].vType of
            vtString:
                case codes[i].vString^ of
                    'fg', 'bg': begin
                        if _catch then flush_color;
                        _param += codes[i].vString^ + ansi_separator;;
                        _catch := true;
                    end
                else begin
                    if _catch then begin
                        _catch := false;
                        flush_color;
                    end;
                    _param += codes[i].vString^ + ansi_separator;
                end;
                end;

            vtInteger:
                if _catch then begin
                    _color[_clind] := codes[i].vInteger;
                    Inc(_clind);
                end;
        else
            exit;
        end;
    end;
    codify := ansi_begin + _param + ansi_end;
end;


function ansify(codes : array of const; const msg : string) : string;
begin
    ansify := codify(codes) + msg + codify([reset]);
end;

end.

File: ocaml/escape.ml


type ansiCode =
  | Reset
  | Bold
  | Faint
  | Italic
  | Underline
  | Blink_Slow
  | Blink_Rapid
  | Reverse
  | Conceal
  | Crossout
  | Font_Primary
  | Font_Alternate_1
  | Font_Alternate_2
  | Font_Alternate_3
  | Font_Alternate_4
  | Font_Alternate_5
  | Font_Alternate_6
  | Font_Alternate_7
  | Font_Alternate_8
  | Font_Alternate_9
  | Fraktur
  | Bold_Off
  | Underline_Double
  | Color_Normal
  | Italic_Off
  | Fraktur_Off
  | Underline_Off
  | Blink_Off
  | Inverse_Off
  | Reveal
  | Crossout_Off
  | Fg_Black
  | Fg_Red
  | Fg_Green
  | Fg_Yellow
  | Fg_Blue
  | Fg_Magenta
  | Fg_Cyan
  | Fg_White
  | Fg  of int
  | Fgx of int * int * int
  | Fg_Default
  | Bg_Black
  | Bg_Red
  | Bg_Green
  | Bg_Yellow
  | Bg_Blue
  | Bg_Magenta
  | Bg_Cyan
  | Bg_White
  | Bg  of int
  | Bgx of int * int * int
  | Bg_Default
  | Frame
  | Encircle
  | Overline
  | Frame_Off
  | Encircle_Off
  | Overline_Off
  | Ideogram_Underline
  | Ideogram_Underline_Double
  | Ideogram_Overline
  | Ideogram_Overline_Double
  | Ideogram_Stressmark
  | Ideogram_Off
  | Fg_Bright_Black
  | Fg_Bright_Red
  | Fg_Bright_Green
  | Fg_Bright_Yellow
  | Fg_Bright_Blue
  | Fg_Bright_Magenta
  | Fg_Bright_Cyan
  | Fg_Bright_White
  | Bg_Bright_Black
  | Bg_Bright_Red
  | Bg_Bright_Green
  | Bg_Bright_Yellow
  | Bg_Bright_Blue
  | Bg_Bright_Magenta
  | Bg_Bright_Cyan
  | Bg_Bright_White

let stringify = function
  | Reset                     -> "0"
  | Bold                      -> "1"
  | Faint                     -> "2"
  | Italic                    -> "3"
  | Underline                 -> "4"
  | Blink_Slow                -> "5"
  | Blink_Rapid               -> "6"
  | Reverse                   -> "7"
  | Conceal                   -> "8"
  | Crossout                  -> "9"
  | Font_Primary              -> "10"
  | Font_Alternate_1          -> "11"
  | Font_Alternate_2          -> "12"
  | Font_Alternate_3          -> "13"
  | Font_Alternate_4          -> "14"
  | Font_Alternate_5          -> "15"
  | Font_Alternate_6          -> "16"
  | Font_Alternate_7          -> "17"
  | Font_Alternate_8          -> "18"
  | Font_Alternate_9          -> "19"
  | Fraktur                   -> "20"
  | Bold_Off                  -> "21"
  | Underline_Double          -> "21"
  | Color_Normal              -> "22"
  | Italic_Off                -> "23"
  | Fraktur_Off               -> "23"
  | Underline_Off             -> "24"
  | Blink_Off                 -> "25"
  | Inverse_Off               -> "27"
  | Reveal                    -> "28"
  | Crossout_Off              -> "29"
  | Fg_Black                  -> "30"
  | Fg_Red                    -> "31"
  | Fg_Green                  -> "32"
  | Fg_Yellow                 -> "33"
  | Fg_Blue                   -> "34"
  | Fg_Magenta                -> "35"
  | Fg_Cyan                   -> "36"
  | Fg_White                  -> "37"
  | Fg (color)                -> "38;5;" ^ string_of_int color
  | Fgx (r, g, b)             -> List.fold_left (fun s c -> s ^ ";" ^ string_of_int c) "38;2" [r; g; b]
  | Fg_Default                -> "39"
  | Bg_Black                  -> "40"
  | Bg_Red                    -> "41"
  | Bg_Green                  -> "42"
  | Bg_Yellow                 -> "43"
  | Bg_Blue                   -> "44"
  | Bg_Magenta                -> "45"
  | Bg_Cyan                   -> "46"
  | Bg_White                  -> "47"
  | Bg (color)                -> "48;5;" ^ string_of_int color
  | Bgx (r, g, b)             -> List.fold_left (fun s c -> s ^ ";" ^ string_of_int c) "48;2" [r; g; b]
  | Bg_Default                -> "49"
  | Frame                     -> "51"
  | Encircle                  -> "52"
  | Overline                  -> "53"
  | Frame_Off                 -> "54"
  | Encircle_Off              -> "54"
  | Overline_Off              -> "55"
  | Ideogram_Underline        -> "60"
  | Ideogram_Underline_Double -> "61"
  | Ideogram_Overline         -> "62"
  | Ideogram_Overline_Double  -> "63"
  | Ideogram_Stressmark       -> "64"
  | Ideogram_Off              -> "65"
  | Fg_Bright_Black           -> "90"
  | Fg_Bright_Red             -> "91"
  | Fg_Bright_Green           -> "92"
  | Fg_Bright_Yellow          -> "93"
  | Fg_Bright_Blue            -> "94"
  | Fg_Bright_Magenta         -> "95"
  | Fg_Bright_Cyan            -> "96"
  | Fg_Bright_White           -> "97"
  | Bg_Bright_Black           -> "100"
  | Bg_Bright_Red             -> "101"
  | Bg_Bright_Green           -> "102"
  | Bg_Bright_Yellow          -> "103"
  | Bg_Bright_Blue            -> "104"
  | Bg_Bright_Magenta         -> "105"
  | Bg_Bright_Cyan            -> "106"
  | Bg_Bright_White           -> "107"

let (<|) f g x = f (g x)
let (<<) f x = f x

let codify =
  (fun s -> String.sub s 0 (String.length s - 1) ^ "m") <|
  List.fold_left (fun s c -> s ^ stringify c ^ ";") "\x1b["

let ansify codes msg =
  let ansi_codes = codify codes in
  let ansi_codes = String.sub ansi_codes 0 (String.length ansi_codes - 1) ^ "m" in
    ansi_codes ^ msg ^ codify [Reset]


let () =
  let () = ansify [Bold; Fg_Red] "Fatality!" |> print_endline in
  let () = ansify [Bold; Underline] ("Cor " ^ codify [Fgx (34, 238, 177)] ^ "bonita") |> print_endline in
    ()
